### PR TITLE
52 as a dev i want the information extracted by the performance assessmentpy script to be aggregated in a human readable report

### LINKED ISSUE #52: As a dev, I want the information extracted by the performance_assessment.py script to be aggregated in a human readable report.
## Description  

**Context**
Our initial `performance_assessment` script currently generates raw data on the system's performance and compiles it into a CSV report. To be useful to our team, this information needs to be transformed into a human-readable format.

**Problem Statement**
To derive meaningful insights from the raw data generated by our `performance_assessment` script, we need to compile the results in the CSV report into an aggregated and understandable format. For now, we will implement a basic aggregation method.

For each label, we will have:
- The average Levenshtein distance score for each field.
- The ratio of fields that are considered to have a passing score.
- The ratio of missing fields from the pipeline output compared to expected results.

**Acceptance Criteria**
To consider this issue resolved, the following criteria need to be met:

1. **Data Aggregation:**
   - [ ] The `performance_assessment` script must aggregate raw data and calculate the average Levenshtein distance score for each field across all labels.
   - [ ] The script must calculate and include in the report the ratio of fields that meet or exceed a predefined passing score threshold.
   - [ ] The script should also calculate and provide the ratio of missing fields from the pipeline's output compared to expected results.

2. **Report Generation:**
   - [ ] The aggregated results should include simple visualizations (e.g., bar charts or pie charts) for better comprehension.
   - [ ] The aggregated data must be compiled into a user-friendly, human-readable report format (e.g., PDF or web dashboard).
   - [ ] Ensure that there is a clear definition at the beginning of the report explaining what constitutes a 'passing score' for field accuracy.

### PR DESCRIPTION
**Pull request by Endlessflow**
TLDR Changes:
- Adding a simple `performance_aggregation.py` script that allows the aggregation of the performance report .
- Consolidated all performance related script into a simple `performance_test.py` script for ease of use.

---

I doubt that anyone will ever run the `performance_assessment.py` script which assesses the performance of the pipeline and output the results into a csv file and the `performance_aggregation.py` script that aggregates that data into a markdown document separately.

That being the case, for ease of use, I consolidated everything into a unified `performance_test.py` script.

In case anyone every needs them (I doubt) - I moved the `performance_assessment.py` and `performance_aggregation.py` in the script folder. (still cant make script work from that folder btw).

closes #52

### CODE REVIEW

**Comment by ChromaticPanic** [comment]
do you need to save the csv? Why not skip it and use the pandas dataframes directly in the next steps?

**Comment by ChromaticPanic** [comment]
What error are you getting when trying to run the script?

**Comment by Endlessflow** [comment]
> do you need to save the csv? Why not skip it and use the pandas dataframes directly in the next steps?

initially the idea was to have the ability to snapshot the data about the performance of the system into a csv to give the liberty of being able to aggregate and use that data however we see fit down the line.

but since we are just going to be moving to the DSPy way of doing eval very soon, the idea with this was just to hack a simple aggregation script that reads the csv.

the option to run only the assessment or only the aggregation is still there - but i think most of the time people will be interested in running both back to back so i unified them into 1 script mostly for ease of use.

**Comment by Endlessflow** [comment]
> What error are you getting when trying to run the script?

For some reason if you move any of the scripts into the script folder, they stop working. 
The scripts stop having access to all the dependencies in the pipeline module. 
It's probably something really simple and stupid that I am missing but I can't seem to find the issue.

**Comment by ChromaticPanic** [comment]
> > What error are you getting when trying to run the script?
> 
> For some reason if you move any of the scripts into the script folder, they stop working. The scripts stop having access to all the dependencies in the pipeline module. It's probably something really simple and stupid that I am missing but I can't seem to find the issue.

It has something to do with how python looks for modules. It can't find the pipelines module. I think if you add 
sys.path.append('/path/to/folder/with/modules') or sys.path.append('../')
Then the scripts in the scripts folder will work.

**Comment by Endlessflow** [comment]
> > > What error are you getting when trying to run the script?
> > 
> > 
> > For some reason if you move any of the scripts into the script folder, they stop working. The scripts stop having access to all the dependencies in the pipeline module. It's probably something really simple and stupid that I am missing but I can't seem to find the issue.
> 
> It has something to do with how python looks for modules. It can't find the pipelines module. I think if you add
> 
> `sys.path.append('/path/to/folder/with/modules')` or `sys.path.append('/path/to/folder/pipelines')`
> 
> relative paths should also work `sys.path.append('../')`
> 
> ```python
> sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')) + '/pipelines/')
> ```
> 
> Then the scripts in the scripts folder will work.

OMG thank you

The workaround of using `sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')) + '/pipelines/')` indeed worked. 

But it was unsettling to me that the "proper" way to do this was to append that line to any executable file inside a module.

Nonetheless, since this was a working fix, it allowed me to reverse engineer my way from a working solution to understand why it worked.

From [python's documentation on sys.path](https://docs.python.org/3/library/sys.html#sys.path):

> **sys.path**
> A list of strings that specifies the search path for modules. Initialized from the environment variable [PYTHONPATH](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH), plus an installation-dependent default.
> 
> By default, as initialized upon program startup, a potentially unsafe path is prepended to [sys.path](https://docs.python.org/3/library/sys.html#sys.path) (before the entries inserted as a result of [PYTHONPATH](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH)):
> 
> - python -m module command line: prepend the current working directory.
> - python script.py command line: prepend the script’s directory. If it’s a symbolic link, resolve symbolic links.
> - python -c code and python (REPL) command lines: prepend an empty string, which means the current working directory.

Turns out when you run a Python script, the directory containing the script is added to `sys.path` as the starting point from which Python discovers other modules thus since it thinks `scripts` is the root of the project, it fails to recognize any other modules.

The right way to execute files that are inside directories is to instead use the `python -m module` syntax. This way of executing the script sets the current working directory as the starting point in `sys.path` from which Python discovers modules.

So to call something from the script folder, the solution all along was simply to do:
```
python -m scripts.run_full_performance_assessment
```

This also answers my question about the mystery of why the unit tests in `tests` were able to "find" the dependencies in pipeline without issues - presumably it works because we call those test files with `python -m unittest`.

damn @ChromaticPanic you have no idea how long this problem has been plaguing me in all my python projects.
Thank you for finally allowing me to find peace 😂

**Comment by ChromaticPanic** [comment]
> > > > What error are you getting when trying to run the script?
> > > 
> > > 
> > > For some reason if you move any of the scripts into the script folder, they stop working. The scripts stop having access to all the dependencies in the pipeline module. It's probably something really simple and stupid that I am missing but I can't seem to find the issue.
> > 
> > 
> > It has something to do with how python looks for modules. It can't find the pipelines module. I think if you add
> > `sys.path.append('/path/to/folder/with/modules')` or `sys.path.append('/path/to/folder/pipelines')`
> > relative paths should also work `sys.path.append('../')`
> > ```python
> > sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')) + '/pipelines/')
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then the scripts in the scripts folder will work.
> 
> OMG thank you
> 
> The workaround of using `sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')) + '/pipelines/')` indeed worked.
> 
> But it was unsettling to me that the "proper" way to do this was to append that line to any executable file inside a module.
> 
> Nonetheless, since this was a working fix, it allowed me to reverse engineer my way from a working solution to understand why it worked.
> 
> From [python's documentation on sys.path](https://docs.python.org/3/library/sys.html#sys.path):
> 
> > **sys.path**
> > A list of strings that specifies the search path for modules. Initialized from the environment variable [PYTHONPATH](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH), plus an installation-dependent default.
> > By default, as initialized upon program startup, a potentially unsafe path is prepended to [sys.path](https://docs.python.org/3/library/sys.html#sys.path) (before the entries inserted as a result of [PYTHONPATH](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH)):
> > 
> > * python -m module command line: prepend the current working directory.
> > * python script.py command line: prepend the script’s directory. If it’s a symbolic link, resolve symbolic links.
> > * python -c code and python (REPL) command lines: prepend an empty string, which means the current working directory.
> 
> Turns out when you run a Python script, the directory containing the script is added to `sys.path` as the starting point from which Python discovers other modules thus since it thinks `scripts` is the root of the project, it fails to recognize any other modules.
> 
> The right way to execute files that are inside directories is to instead use the `python -m module` syntax. This way of executing the script sets the current working directory as the starting point in `sys.path` from which Python discovers modules.
> 
> So to call something from the script folder, the solution all along was simply to do:
> 
> ```
> python -m scripts.run_full_performance_assessment
> ```
> 
> This also answers my question about the mystery of why the unit tests in `tests` were able to "find" the dependencies in pipeline without issues - presumably it works because we call those test files with `python -m unittest`.
> 
> damn @ChromaticPanic you have no idea how long this problem has been plaguing me in all my python projects. Thank you for finally allowing me to find peace 😂

That long one guarantees that a relative path is added based on your file location. The other ones are dependent on execution context which can change based on if you are a system or root user or if you are on windows or linux. 

Another strategy I use in my scripts is to explicitly set the current working directory (cwd) before running something like sys.path.append('../')

python -m would probably only work in this configuration where the scripts folder is beside the module folder. not sure how well it would work in more complicated module nesting. But if the module nesting is complicated the project should probably be refactored anyway.

Glad to be of help!

**Review by k-allagbe — APPROVED** [review]
